### PR TITLE
Build time configuration for CSR

### PR DESCRIPTION
**Pull request by james-ctc**
# Background
I'm using the client certificate as a TLS server certificate for MQTT and need to include the required subject alternative name fields in the certificate signing request (CSR).

i.e. subject alternative name can be added with DNS name and/or IPv4 address

# Implementation
By default the subject alternative name extension is not added to the CSR. When CSR_DNS_NAME or CSR_IP_ADDRESS are set using cmake then a subject alternative name extension is created and populated with the supplied values,

e.g.
```
cmake -DCSR_DNS_NAME=install.pionix.de
```

## Possible future enhancement
APIs and methods could be updated to enable run-time configuration

### CODE REVIEW

**Comment by AssemblyJohn** [comment]
I do not think that a compile-time configuration is a good option for this, the interface should be modified to support this for runtime only. We will clog the make with too much info. I think there should be added one more interface function 'gen_csr' whatver that receives a full struct with all the possible params so that in the future we will not be blocked by any other necessary params. On my side this doesn't have a go. @Pietfried what is your take on this?

**Comment by james-ctc** [comment]
Suppose I want to use RSA rather than EC for the CSR. At the moment this is hard coded to CryptoKeyType::EC_prime256v1
If it was configurable (as done for CSR_DNS_NAME) then it would be easy to choose. The only option remaining is to edit evse_security.cpp (or apply a patch to it). 

I feel supporting a cmake define is a much cleaner solution than requiring file edits or patches.

Perhaps there is an intent for info.key_info.key_type to be configurable, but it isn't yet.
Hence could not this be treated in the same way? i.e. a starting point and not the full solution.

This approach doesn't change any method calls or MQTT messages and hence is safe. The defaults leave existing operation untouched.

**Comment by AssemblyJohn** [comment]
That is correct, I've made a commit to fix the issue. We only need to modify the relevant interfaces to use it. Again, I would not hard-code these values in code, even less in the cmake.

**Comment by hikinggrass** [comment]
Yes I agree, moving directly to making this runtime configurable is probably the best way forward

**Comment by james-ctc** [comment]
I'm not keen on run-time configuration since the information in the CSR ***must*** match the information in system setting files for hostapd, systemd network information, and dnsmasq.

In yocto I can configure variables so that all the configuration is consistent and can't be accidentally changed.
I feel runtime configuration should not be used where a change in the setting can break the system.

`EXTRA_OECMAKE += "-DCSR_DNS_NAME=${CHARGER_DNS_NAME} -DCSR_IP_ADDRESS=${HOTSPOT_IPADDRESS}"`

is preferable to applying a patch to the code.

**Comment by hikinggrass** [comment]
> I'm not keen on run-time configuration since the information in the CSR _**must**_ match the information in system setting files for hostapd, systemd network information, and dnsmasq.
> 
> In yocto I can configure variables so that all the configuration is consistent and can't be accidentally changed. I feel runtime configuration should not be used where a change in the setting can break the system.
> 
> `EXTRA_OECMAKE += "-DCSR_DNS_NAME=${CHARGER_DNS_NAME} -DCSR_IP_ADDRESS=${HOTSPOT_IPADDRESS}"`
> 
> is preferable to applying a patch to the code.

With this extra info/additional constraints I would also support (at least the possibility) of making this compile-time configurable. There's an argument to be made for runtime configurable settings during development, but on an embedded system you probably want to limit the possibilities of mis-configuring something

**Comment by AssemblyJohn** [comment]
I've also made some integrated changes for a full CSR option (generate_csr_full). I see they're gone, where did they go? In case they're lost I'll only be able to approve after I'll be able to add them back, since this CSR will only grow in the future and if we're on it I'd prefer to fix it now rather than micro-patch it every 1-2 weeks.

**Review by AssemblyJohn — APPROVED** [review]

**Pull request merged**
